ygen/schemaparse.go

```diff
@@ -17,9 +17,79 @@ package ygen
 import (
 	"bytes"
 	"compress/gzip"
+	"encoding/json"
 	"fmt"
+	"strings"
+
+	"github.com/openconfig/goyang/pkg/yang"
 )
 
+// buildJSON tree takes an input set of modules expressed as a slice of yang.Entry
+// structs, the names of directories within the generated code, and a yang.Entry
+// which corresponds to the fake root and returns a JSON serialised tree of the
+// schema for the set of modules. The JSON document that is returned is always
+// rooted on a yang.Entry which corresponds to the root item, and stores all
+// root-level enties (and their subtrees) within the input module set. All
+// YANG directories are annotated in the output JSON with the name of the type
+// they correspond to in the generated code, and the absolute schema path that
+// the entry corresponds to.
+func buildJSONTree(ms []*yang.Entry, dn map[string]string, fakeroot *yang.Entry) ([]byte, error) {
+	rootEntry := &yang.Entry{
+		Dir:        map[string]*yang.Entry{},
+		Annotation: map[string]interface{}{},
+	}
+	for _, m := range ms {
+		annotateChildren(m, dn)
+		for _, ch := range children(m) {
+			if _, ex := rootEntry.Dir[ch.Name]; ex {
+				return nil, fmt.Errorf("overlapping root children for key %s", ch.Name)
+			}
+			rootEntry.Dir[ch.Name] = ch
+		}
+	}
+
+	if fakeroot != nil {
```

Comment:
Is fakeroot == nil going to be allowed? If so, the function comment should be updated.

Reply:
It just made for a very big CL to change it in this CL too, so I was planning to make that change in a subsequent one. Updated the function comment.